### DISCUSSION TITLE
Comparing GeneratedColumnWithTypeConverter

### DISCUSSION THREAD

**Discussion by zmeid** [simolus3/drift#2018] [Q&A] [answered]
Hi,

I have a column that is converted with TypeConverter:

```
class SourceDtoTypeConverter extends TypeConverter<SourceDto, String> {
  @override
  SourceDto fromSql(String fromDb) {
    return SourceDto.fromJson((jsonDecode(fromDb) as Map<String, dynamic>));
  }

  @override
  String toSql(SourceDto value) {
    return jsonEncode(value);
  }
}
```

**SourceDto:**

```
part 'source_dto.g.dart';

@JsonSerializable()
class SourceDto {
  String id;
  String name;

  SourceDto(this.id, this.name);

  factory SourceDto.fromJson(Map<String, dynamic> json) => _$SourceDtoFromJson(json);

  Map<String, dynamic> toJson() => _$SourceDtoToJson(this);

  /// Currently comparing only with name.
  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is SourceDto && runtimeType == other.runtimeType && name == other.name;

  @override
  int get hashCode => name.hashCode;
}
```

Since I make the ```SourceDto``` comparable by ```name```, I expect Drift to compare it by name in the queries. But, this where statement in the query always returns false:

```
..where(
          (tbl) => tbl.source.isIn(["name1", "name2", "name3"])
```

I also tried ```equals``` instead of ```isIn```, but it still always returns false:

```
..where(
          (tbl) => tbl.source.equals("name1")
```

Am I missing something? Is my approach wrong? How can I filter rows by ```name``` property of SourceDto?

**Comment by simolus3** [comment, marked as answer]
Doing `isIn` will compare to the (json) value in SQL, it will not invoke the type converter. Type converters are only used to convert values to SQL and back to Dart (when reading results). To respect your `==` operator, drift would have to map every intermediate row back to Dart, apply the type converter and use that as a filter. That would be pretty inefficient and defeat the point of using a SQL engine.

As sqlite3 has builtin json support, you can compare names like this though:

```dart
import 'package:drift/extensions/json1.dart';

..where(
          (tbl) => tbl.source.jsonExtract<String>(r'$.name').isIn(["name1", "name2", "name3"])
```

For details, see [`json_extract`'s documentation](https://www.sqlite.org/json1.html#jex).
